src/microbe_stage/ColonyCompoundBag.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+
+[UseThriveSerializer]
+public class ColonyCompoundBag : ICompoundStorage
+{
+    public ColonyCompoundBag()
+    {
+    }
+
+    public ColonyCompoundBag(MicrobeColony colony)
+    {
+        Colony = colony;
+    }
+
+    [JsonProperty]
+    public MicrobeColony Colony { get; set; }
+
+    public float Capacity => GetCompoundBags().Sum(p => p.Capacity);
+
+    /// <summary>
+    ///   Distributes the amount above average compounds to the cells below average compounds.
+    /// </summary>
+    public void DistributeCompoundSurplus()
+    {
+        var bags = GetCompoundBags().ToList();
+        using var compounds = GetEnumerator();
+
+        while (compounds.MoveNext())
+        {
+            var compound = compounds.Current;
+            if (!compound.Key.CanBeDistributed)
+                continue;
+
+            var average = compound.Value / bags.Count;
+
+            foreach (var bag in bags)
+            {
+                var surplus = bag.GetCompoundAmount(compound.Key) - average;
+                if (surplus > 0)
+                    bag.TakeCompound(compound.Key, surplus);
+                else
+                    bag.AddCompound(compound.Key, -surplus);
+            }
+        }
+    }
+
+    public IEnumerator<KeyValuePair<Compound, float>> GetEnumerator()
+    {
+        return GetCompoundBags()
+            .SelectMany(p => p.Compounds)
+            .GroupBy(p => p.Key)
+            .Select(p => new KeyValuePair<Compound, float>(p.Key, p.Sum(x => x.Value)))
+            .GetEnumerator();
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+
+    public float GetCompoundAmount(Compound compound)
+    {
+        return GetCompoundBags().Sum(p => p.GetCompoundAmount(compound));
+    }
+
+    public float TakeCompound(Compound compound, float amount)
+    {
+        // bags that compound can be taken from
+        var remainingBags = GetCompoundBags().ToList();
+
+        while (amount > MathUtils.EPSILON && remainingBags.Any())
```

Comment:
I use a while loop, because I modify the `remaindingBags` list inside the loop. ForEach-loops do not allow the modification of the list inside the iteration block.

Reply:
Did I miss something? You are just modifying the compound amounts. You can't modify the foreach variable, but if it is a reference you can modify that referred-to object all you want.

Reply:
The logic currently works like this:
I have a list of bags I can potentially take compounds from.
The loop takes the first bag in the list and takes compounds. If the bag is dry it gets removed from the list and another bag is the first in the list.

I could rewrite this code to use an actual Stack or Queue and a variable with the current bag to take compounds from.

Reply:
Why do you need that? You recreate the remainingBags list each time this method is called. What point there is to them? You could just as well do a foreach starting with the first bag, and breaking once you have taken enough compounds.

Reply:
Now that I think about it... I have no idea why I did it that way :D